File: Python3/601 - 800/693 Binary Number with Alternating Bits.py
# 36ms, 100.0%
class Solution:
    def hasAlternatingBits(self, n):
        """
        :type n: int
        :rtype: bool
        """
        curMax = 2**int(math.log2(n) + 1)
        while n > 1:
            n = curMax - n - 1
            curMax >>= 1
        return n == 1 and curMax == 2
    

File: 601 - 800/693. Binary Number with Alternating Bits.cc
// 0ms, 100.0%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    bool hasAlternatingBits(int n) {
        long long curMax = pow(2, 1 + int(log(n) / log(2)));
        while (n > 1) {
            n = curMax - n - 1;
            curMax >>= 1;
        }
        return n == 1 && curMax == 2;
    }
};
